File: src/godot/hap_player.cpp
#include "hap_player.h"

#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/core/error_macros.hpp>

namespace godot {

void HapPlayer::_bind_methods() {
  ClassDB::bind_method(D_METHOD("set_stream", "stream"), &HapPlayer::set_stream);
  ClassDB::bind_method(D_METHOD("get_stream"), &HapPlayer::get_stream);

  ClassDB::bind_method(D_METHOD("set_loop", "loop"), &HapPlayer::set_loop);
  ClassDB::bind_method(D_METHOD("is_loop"), &HapPlayer::is_loop);

  ClassDB::bind_method(D_METHOD("set_playback_speed", "speed"),
                       &HapPlayer::set_playback_speed);
  ClassDB::bind_method(D_METHOD("get_playback_speed"),
                       &HapPlayer::get_playback_speed);

  ClassDB::bind_method(D_METHOD("set_autoplay", "autoplay"),
                       &HapPlayer::set_autoplay);
  ClassDB::bind_method(D_METHOD("is_autoplay"), &HapPlayer::is_autoplay);

  ClassDB::bind_method(D_METHOD("set_stream_position", "position"),
                       &HapPlayer::set_stream_position);
  ClassDB::bind_method(D_METHOD("get_stream_position"),
                       &HapPlayer::get_stream_position);

  ClassDB::bind_method(D_METHOD("set_paused", "paused"), &HapPlayer::set_paused);
  ClassDB::bind_method(D_METHOD("is_paused"), &HapPlayer::is_paused);

  ClassDB::bind_method(D_METHOD("play"), &HapPlayer::play);
  ClassDB::bind_method(D_METHOD("pause"), &HapPlayer::pause);
  ClassDB::bind_method(D_METHOD("stop"), &HapPlayer::stop);
  ClassDB::bind_method(D_METHOD("step_frame", "n"), &HapPlayer::step_frame);

  ClassDB::bind_method(D_METHOD("get_texture"), &HapPlayer::get_texture);

  ClassDB::bind_method(D_METHOD("get_frame_rate"), &HapPlayer::get_frame_rate);
  ClassDB::bind_method(D_METHOD("get_width"), &HapPlayer::get_width);
  ClassDB::bind_method(D_METHOD("get_height"), &HapPlayer::get_height);
  ClassDB::bind_method(D_METHOD("get_duration"), &HapPlayer::get_duration);
  ClassDB::bind_method(D_METHOD("get_frame_count"), &HapPlayer::get_frame_count);

  ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "stream", PROPERTY_HINT_RESOURCE_TYPE,
                            "HapVideoStream"),
               "set_stream", "get_stream");
  ADD_PROPERTY(PropertyInfo(Variant::BOOL, "loop"), "set_loop", "is_loop");
  ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "playback_speed"),
               "set_playback_speed", "get_playback_speed");
  ADD_PROPERTY(PropertyInfo(Variant::BOOL, "autoplay"), "set_autoplay",
               "is_autoplay");
  ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "stream_position"),
               "set_stream_position", "get_stream_position");
  ADD_PROPERTY(PropertyInfo(Variant::BOOL, "paused"), "set_paused", "is_paused");

  // Read-only metadata, valid after `opened` -- no setter, so scripts
  // read them as plain properties (player.frame_rate, not
  // player.get_frame_rate()) via dot access, matching the spec's
  // property list.
  ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "frame_rate", PROPERTY_HINT_NONE, "",
                            PROPERTY_USAGE_READ_ONLY),
               "", "get_frame_rate");
  ADD_PROPERTY(PropertyInfo(Variant::INT, "width", PROPERTY_HINT_NONE, "",
                            PROPERTY_USAGE_READ_ONLY),
               "", "get_width");
  ADD_PROPERTY(PropertyInfo(Variant::INT, "height", PROPERTY_HINT_NONE, "",
                            PROPERTY_USAGE_READ_ONLY),
               "", "get_height");
  ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "duration", PROPERTY_HINT_NONE, "",
                            PROPERTY_USAGE_READ_ONLY),
               "", "get_duration");
  ADD_PROPERTY(PropertyInfo(Variant::INT, "frame_count", PROPERTY_HINT_NONE, "",
                            PROPERTY_USAGE_READ_ONLY),
               "", "get_frame_count");

  ADD_SIGNAL(MethodInfo("opened"));
  ADD_SIGNAL(MethodInfo("playback_completed"));
  ADD_SIGNAL(MethodInfo("playback_looped"));
  ADD_SIGNAL(MethodInfo("error_occurred", PropertyInfo(Variant::STRING, "message")));
}

void HapPlayer::set_stream(const Ref<HapVideoStream> &p_stream) {
  stream_ = p_stream;
  playback_.unref();

  error_fired_ = false;
  active_ = false;
  paused_ = false;
  position_ = 0.0;
  needs_retarget_ = true;
  last_direction_forward_ = true;

  if (stream_.is_null())
    return;

  Ref<HapVideoStreamPlayback> pb;
  pb.instantiate();
  if (!pb->open(stream_->get_file())) {
    ERR_PRINT("HapPlayer: failed to open stream: " + stream_->get_file());
    return;
  }
  playback_ = pb;
}

void HapPlayer::set_stream_position(double p_position) {
  if (p_position < 0.0)
    p_position = 0.0;

  if (playback_.is_valid() && playback_->is_ready()) {
    double duration = playback_->_get_length();
    if (p_position > duration)
      p_position = duration;
  }

  position_ = p_position;
  needs_retarget_ = true;

  if (playback_.is_valid() && playback_->is_ready()) {
    bool forward = playback_speed_ >= 0.0;
    playback_->advance_to_frame(frame_from_position(position_), forward, true);
    last_direction_forward_ = forward;
    needs_retarget_ = false;
  }
}

void HapPlayer::play() {
  active_ = true;
  paused_ = false;
  needs_retarget_ = true;
}

void HapPlayer::pause() { paused_ = true; }

void HapPlayer::stop() {
  active_ = false;
  paused_ = false;
  position_ = 0.0;
  needs_retarget_ = true;
  last_direction_forward_ = true;

  if (playback_.is_valid() && playback_->is_ready()) {
    playback_->advance_to_frame(0, /*forward=*/true, /*retarget=*/true);
    needs_retarget_ = false;
  }
}

void HapPlayer::step_frame(int32_t n) {
  if (playback_.is_null() || !playback_->is_ready())
    return;

  // Pause-gated: auto-pause if currently advancing, then step exactly
  // n frames via a priority seek.
  if (active_ && !paused_)
    paused_ = true;

  int32_t frame_count = playback_->get_frame_count();
  int32_t current_frame = static_cast<int32_t>(frame_from_position(position_));
  int32_t target = current_frame + n;
  if (target < 0)
    target = 0;
  if (frame_count > 0 && target >= frame_count)
    target = frame_count - 1;

  bool forward = n >= 0;
  double frame_rate = playback_->get_frame_rate();
  position_ = frame_rate > 0.0 ? target / frame_rate : 0.0;

  playback_->advance_to_frame(static_cast<uint32_t>(target), forward, true);
  last_direction_forward_ = forward;
  needs_retarget_ = false;
}

Ref<Texture2D> HapPlayer::get_texture() const {
  if (playback_.is_null())
    return Ref<Texture2D>();
  return playback_->_get_texture();
}

double HapPlayer::get_frame_rate() const {
  return playback_.is_valid() ? playback_->get_frame_rate() : 0.0;
}

int32_t HapPlayer::get_width() const {
  return playback_.is_valid() ? playback_->get_width() : 0;
}

int32_t HapPlayer::get_height() const {
  return playback_.is_valid() ? playback_->get_height() : 0;
}

double HapPlayer::get_duration() const {
  return playback_.is_valid() ? playback_->_get_length() : 0.0;
}

int32_t HapPlayer::get_frame_count() const {
  return playback_.is_valid() ? playback_->get_frame_count() : 0;
}

uint32_t HapPlayer::frame_from_position(double p_position) const {
  if (playback_.is_null() || !playback_->is_ready())
    return 0;

  double frame_rate = playback_->get_frame_rate();
  int32_t frame_count = playback_->get_frame_count();
  uint32_t frame =
      frame_rate > 0.0 ? static_cast<uint32_t>(p_position * frame_rate) : 0;
  if (frame_count > 0 && frame >= static_cast<uint32_t>(frame_count))
    frame = static_cast<uint32_t>(frame_count) - 1;
  return frame;
}

void HapPlayer::_ready() { set_process(true); }

void HapPlayer::_process(double p_delta) {
  if (playback_.is_null())
    return;

  bool was_ready = playback_->is_ready();
  bool now_ready = playback_->poll_ready();

  if (now_ready && !was_ready) {
    // Materialize the current (possibly pre-set) position's frame
    // immediately, so get_texture() is valid the instant `opened`
    // fires -- before any handler calls play().
    bool forward = playback_speed_ >= 0.0;
    playback_->advance_to_frame(frame_from_position(position_), forward, true);
    last_direction_forward_ = forward;
    needs_retarget_ = false;

    emit_signal("opened");
    if (autoplay_)
      play();
  }

  if (playback_->has_failed()) {
    if (!error_fired_) {
      error_fired_ = true;
      emit_signal("error_occurred", playback_->get_error());
    }
    return;
  }

  if (!now_ready || !active_ || paused_)
    return;

  hap::core::PumpStepResult step = hap::core::pump_step(
      position_, p_delta, playback_speed_, loop_, playback_->_get_length(),
      needs_retarget_, last_direction_forward_);

  position_ = step.position;
  needs_retarget_ = false;
  last_direction_forward_ = step.forward;
  if (step.completed)
    active_ = false;

  playback_->advance_to_frame(frame_from_position(position_), step.forward,
                              step.retarget);

  if (step.looped)
    emit_signal("playback_looped");
  if (step.completed)
    emit_signal("playback_completed");
}

} // namespace godot

File: src/godot/hap_player.zig
//! hap_player.zig — the HapPlayer Control node.
//!
//! Power-user layer: a Control node that owns the authoritative Godot-free
//! PlaybackController and drives a thin HapVideoStreamPlayback adapter.
//!
//! The GDScript surface (methods/properties/signals) is exercised
//! end-to-end by project/demo/hap_player_demo.gd: stream, loop, autoplay, opened,
//! playback_completed, error_occurred, frame_count, stream_position,
//! frame_rate, playback_speed, paused, get_texture(), play(), pause(),
//! step_frame(n).
//!
const HapPlayer = @This();

const std = @import("std");
const Allocator = std.mem.Allocator;

const godot = @import("godot");
const Registry = godot.extension.Registry;
const Control = godot.class.Control;
const Texture2d = godot.class.Texture2d;
const String = godot.builtin.String;
const PropertyUsageFlags = godot.global.PropertyUsageFlags;

// Core types come through the "core" named module (build.zig-wired) so they
// match the module instance the rest of the extension hands around.
const core = @import("core");
const PlaybackController = core.playback_controller.PlaybackController;

const HapVideoStream = @import("hap_video_stream.zig");
const HapVideoStreamPlayback = @import("hap_video_stream_playback.zig");
const ref = @import("ref.zig");

const log = std.log.scoped(.hap_player);

/// Usage for the read-only metadata properties below (frame_rate, width,
/// height, duration, frame_count). Read-only: storage and editor visibility
/// both off.
const read_only_usage: PropertyUsageFlags = .{
    .property_usage_storage = false,
    .property_usage_editor = false,
    .property_usage_read_only = true,
};

pub fn register(r: *Registry) void {
    const class = r.createClass(HapPlayer, r.allocator, .auto);

    class.addMethod("play", .auto);
    class.addMethod("pause", .auto);
    class.addMethod("stop", .auto);
    class.addMethod("step_frame", .auto);
    class.addMethod("get_texture", .auto);

    class.addProperty("stream", .{
        .hint = .property_hint_resource_type,
        .hint_string = String.fromLatin1("HapVideoStream"),
    });

    // "loop"/"paused"/"autoplay" use is_x()/set_x() naming (Godot's bool
    // convention), which gdzig's property auto-detect (getX/setX only)
    // can't find on its own -- register the methods explicitly and wire
    // them to the property by pointer.
    const is_loop = class.createMethod("is_loop", .auto);
    const set_loop = class.createMethod("set_loop", .auto);
    class.addProperty("loop", .{ .getter = .{ .method = is_loop }, .setter = .{ .method = set_loop } });

    class.addProperty("playback_speed", .auto);

    const is_autoplay = class.createMethod("is_autoplay", .auto);
    const set_autoplay = class.createMethod("set_autoplay", .auto);
    class.addProperty("autoplay", .{ .getter = .{ .method = is_autoplay }, .setter = .{ .method = set_autoplay } });

    class.addProperty("stream_position", .auto);

    const is_paused = class.createMethod("is_paused", .auto);
    const set_paused = class.createMethod("set_paused", .auto);
    class.addProperty("paused", .{ .getter = .{ .method = is_paused }, .setter = .{ .method = set_paused } });

    // Read-only metadata, valid after `opened` -- no setter.
    class.addProperty("frame_rate", .{ .setter = .none, .usage = read_only_usage });
    class.addProperty("width", .{ .setter = .none, .usage = read_only_usage });
    class.addProperty("height", .{ .setter = .none, .usage = read_only_usage });
    class.addProperty("duration", .{ .setter = .none, .usage = read_only_usage });
    class.addProperty("frame_count", .{ .setter = .none, .usage = read_only_usage });

    class.addSignal(Opened);
    class.addSignal(PlaybackCompleted);
    class.addSignal(PlaybackLooped);
    class.addSignal(ErrorOccurred);
}

pub fn unregister(r: *Registry) void {
    r.removeClass(HapPlayer);
}

pub const Opened = struct {};
pub const PlaybackCompleted = struct {};
pub const PlaybackLooped = struct {};
pub const ErrorOccurred = struct { message: String };

allocator: Allocator,
base: *Control,

stream: ?*HapVideoStream = null,
playback: ?*HapVideoStreamPlayback = null,
controller: PlaybackController = .init(),

// A replacement remains private until its asynchronous open and GPU setup
// both succeed. It uses a separate controller so candidate initialization
// cannot reconfigure the still-active playback's clock.
pending_stream: ?*HapVideoStream = null,
pending_playback: ?*HapVideoStreamPlayback = null,
pending_controller: PlaybackController = .init(),
autoplay: bool = false,

error_fired: bool = false,

pub fn create(allocator: *Allocator) !*HapPlayer {
    const self = try allocator.create(HapPlayer);
    self.* = .{
        .allocator = allocator.*,
        .base = .init(),
    };
    self.base.setInstance(HapPlayer, self);
    return self;
}

pub fn destroy(self: *HapPlayer, allocator: *Allocator) void {
    self.clearPendingStream();
    self.clearStream();
    self.base.destroy();
    allocator.destroy(self);
}

pub fn setStream(self: *HapPlayer, new_stream: ?*HapVideoStream) void {
    // A property get/set round trip can pass our own stored resource back to
    // us. Keep the active adapter intact rather than releasing the only
    // persistent reference before we retain the replacement.
    if (new_stream != null and self.pending_stream == new_stream) return;
    if (self.stream == new_stream) {
        self.clearPendingStream();
        return;
    }

    const stream = new_stream orelse {
        self.clearPendingStream();
        self.clearStream();
        return;
    };

    // There can be only one unpublished candidate. Cancelling an older open
    // may wait for its already-running stream job, but never disturbs the
    // active playback pair.
    self.clearPendingStream();

    self.pending_controller = self.controller;
    self.pending_controller.resetStream();

    // Take our own persistent share: `new_stream` is a borrowed pointer
    // for the duration of this setter call only (matches every other
    // Object-typed method argument here), but we're storing it beyond
    // that, so we need an independent owning reference.
    _ = stream.base.reference();

    const pb = HapVideoStreamPlayback.createBorrowed(&self.allocator, &self.pending_controller) catch {
        ref.releaseAndDestroy(HapVideoStream, stream, &self.allocator);
        self.emitStreamError("Failed to create playback");
        return;
    };

    var file = stream.getFile();
    defer file.deinit();
    if (!pb.open(file)) {
        var buf: [1024]u8 = undefined;
        log.err("failed to open stream: {s}", .{file.toUtf8Buf(&buf)});
        ref.releaseAndDestroy(HapVideoStreamPlayback, pb, &self.allocator);
        ref.releaseAndDestroy(HapVideoStream, stream, &self.allocator);
        self.emitStreamError("Failed to start stream open");
        return;
    }

    self.pending_playback = pb;
    self.pending_stream = stream;
}

fn clearPendingStream(self: *HapPlayer) void {
    if (self.pending_playback) |pb| {
        self.pending_playback = null;
        ref.releaseAndDestroy(HapVideoStreamPlayback, pb, &self.allocator);
    }
    if (self.pending_stream) |stream| {
        self.pending_stream = null;
        ref.releaseAndDestroy(HapVideoStream, stream, &self.allocator);
    }
    self.pending_controller = .init();
}

fn clearStream(self: *HapPlayer) void {
    if (self.playback) |pb| {
        self.playback = null;
        ref.releaseAndDestroy(HapVideoStreamPlayback, pb, &self.allocator);
    }
    if (self.stream) |stream| {
        self.stream = null;
        ref.releaseAndDestroy(HapVideoStream, stream, &self.allocator);
    }
    self.error_fired = false;
    self.controller.resetStream();
}

fn emitStreamError(self: *HapPlayer, text: []const u8) void {
    var message = String.fromLatin1(text);
    defer message.deinit();
    self.base.emit(ErrorOccurred, .{ .message = message }) catch {};
}

pub fn getStream(self: *HapPlayer) ?*HapVideoStream {
    return self.pending_stream orelse self.stream;
}

/// Public controls target the selected replacement while it is opening. The
/// old playback remains a visual placeholder and keeps its existing state;
/// if the candidate succeeds these pre-open controls are promoted with it,
/// and if it fails the untouched active state becomes authoritative again.
fn selectedController(self: *HapPlayer) *PlaybackController {
    return if (self.pending_stream != null) &self.pending_controller else &self.controller;
}

fn selectedPlayback(self: *HapPlayer) ?*HapVideoStreamPlayback {
    return self.pending_playback orelse self.playback;
}

pub fn isLoop(self: *HapPlayer) bool {
    return self.selectedController().isLooping();
}
pub fn setLoop(self: *HapPlayer, value: bool) void {
    self.selectedController().setLoop(value);
}

pub fn getPlaybackSpeed(self: *HapPlayer) f64 {
    return self.selectedController().playbackSpeed();
}
pub fn setPlaybackSpeed(self: *HapPlayer, value: f64) void {
    self.selectedController().setPlaybackSpeed(value);
}

pub fn isAutoplay(self: *HapPlayer) bool {
    return self.autoplay;
}
pub fn setAutoplay(self: *HapPlayer, value: bool) void {
    self.autoplay = value;
}

pub fn getStreamPosition(self: *HapPlayer) f64 {
    return self.selectedController().position();
}
pub fn setStreamPosition(self: *HapPlayer, position: f64) void {
    const request = self.selectedController().seek(position);
    if (self.selectedPlayback()) |pb| {
        if (pb.isReady()) pb.advance(request);
    }
}

pub fn isPaused(self: *HapPlayer) bool {
    return self.selectedController().isPaused();
}
pub fn setPaused(self: *HapPlayer, value: bool) void {
    self.selectedController().setPaused(value);
}

/// Start (or resume) playback from the current stream_position, in the
/// direction implied by playback_speed's sign. Does not reset position --
/// resuming after pause/completion continues from where playback left off.
pub fn play(self: *HapPlayer) void {
    self.selectedController().play();
}

/// Freeze the pump in place; texture and stream_position are retained.
/// Equivalent to set_paused(true).
pub fn pause(self: *HapPlayer) void {
    self.selectedController().pause();
}

/// Halt playback and reset stream_position to 0.
pub fn stop(self: *HapPlayer) void {
    const request = self.selectedController().stop();
    if (self.selectedPlayback()) |pb| {
        if (pb.isReady()) pb.advance(request);
    }
}

/// Pause-gated: auto-pauses if playing, then moves exactly n frames (n may
/// be negative) via a priority seek. `n` is `i64`: gdzig bound-method
/// marshalling supports only i64 integers; GDScript int is 64-bit anyway.
pub fn stepFrame(self: *HapPlayer, n: i64) void {
    const pb = self.playback orelse return;
    if (!pb.isReady()) return;
    if (self.controller.stepFrame(n)) |request| pb.advance(request);
}

pub fn getTexture(self: *HapPlayer) ?*Texture2d {
    const pb = self.playback orelse return null;
    return pb.texture();
}

// Read-only metadata, valid after the `opened` signal. `i64` return types
// for the same reason as step_frame's `n` above.
pub fn getFrameRate(self: *HapPlayer) f64 {
    return if (self.selectedPlayback()) |pb| pb.getFrameRate() else 0.0;
}
pub fn getWidth(self: *HapPlayer) i64 {
    return if (self.selectedPlayback()) |pb| @intCast(pb.getWidth()) else 0;
}
pub fn getHeight(self: *HapPlayer) i64 {
    return if (self.selectedPlayback()) |pb| @intCast(pb.getHeight()) else 0;
}
pub fn getDuration(self: *HapPlayer) f64 {
    return self.selectedController().duration();
}
pub fn getFrameCount(self: *HapPlayer) i64 {
    return if (self.selectedPlayback()) |pb| @intCast(pb.getFrameCount()) else 0;
}

pub fn _ready(self: *HapPlayer) void {
    self.base.setProcess(true);
}

pub fn _process(self: *HapPlayer, delta: f64) void {
    self.pollPendingStream();

    const pb = self.playback orelse return;

    const was_ready = pb.isReady();
    const now_ready = pb.pollReady();

    if (now_ready and !was_ready) {
        // Materialize the current (possibly pre-set) position's frame
        // immediately, so getTexture() is valid the instant `opened` fires
        // -- before any handler calls play().
        pb.advance(self.controller.frameRequest(true));

        self.base.emit(Opened, .{}) catch {};
        if (self.autoplay) self.play();
    }

    if (pb.hasFailed()) {
        if (!self.error_fired) {
            self.error_fired = true;
            var message = pb.getError();
            defer message.deinit();
            self.base.emit(ErrorOccurred, .{ .message = message }) catch {};
        }
        return;
    }

    if (!now_ready) return;

    const step = pb.updateController(delta);

    if (step.looped) self.base.emit(PlaybackLooped, .{}) catch {};
    if (step.completed) self.base.emit(PlaybackCompleted, .{}) catch {};
}

/// Advance an unpublished replacement through async open and GPU setup. A
/// failure releases only the candidate; success swaps the complete pair and
/// rebinds its controller to the player's stable active-controller address.
fn pollPendingStream(self: *HapPlayer) void {
    const pb = self.pending_playback orelse return;

    const ready = pb.pollReady();
    if (pb.hasFailed()) {
        var message = pb.getError();
        defer message.deinit();
        self.clearPendingStream();
        self.base.emit(ErrorOccurred, .{ .message = message }) catch {};
        return;
    }
    if (!ready) return;

    const stream = self.pending_stream.?;
    const controller = self.pending_controller;
    self.pending_playback = null;
    self.pending_stream = null;
    self.pending_controller = .init();

    self.clearStream();
    self.controller = controller;
    pb.rebindController(&self.controller);
    self.playback = pb;
    self.stream = stream;
    self.error_fired = false;

    self.base.emit(Opened, .{}) catch {};
    if (self.autoplay) self.play();
}
